First pass at using bluesky-widgets to generate thumbnails.

danielballan opened the pull request:
This uses a yet-to-be-merged branch on bluesky-widgets: https://github.com/bluesky/bluesky-widgets/pull/80

Aside from the refactor to use bluesky-widgets, it will result in generating a directory of thumbnails per run rather than one thumbnail per run.

I think this is close, but it doesn't _quite_ work yet It seems that the thumbnail generation code is trying to operate on a `h5py.Dataset` in an `h5py.File` that has already been closed. See detail box for full traceback from pytest. This is primarily a note-to-future-self, but if anything pops out at you @dylanmcreynolds I wouldn't refuse the help. :- D

<details>

```
_________________________________________________________________________________ test_hdf5_mapped_ingestor __________________________________________________________________________________

sample_file = <HDF5 file "test.hdf5" (mode r)>, tmp_path = PosixPath('/tmp/pytest-of-dallan/pytest-30/test_hdf5_mapped_ingestor0')

    def test_hdf5_mapped_ingestor(sample_file, tmp_path):
        ingestor = MappedHD5Ingestor(Mapping(**mapping_dict), sample_file, "test_root", thumbs_root=tmp_path)
        run_cache = SingleRunCache()
        descriptors = []
        result_events = []
        result_datums = []
        start_found = False
        stop_found = False
        run_uid = ""
>       for name, doc in ingestor.generate_docstream():

splash_ingest/tests/test_mapping_ingestor.py:115: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
splash_ingest/ingestors.py:194: in generate_docstream
    self._build_thumbnail(start_doc["uid"], self._thumbs_root, bluesky_run)
splash_ingest/ingestors.py:200: in _build_thumbnail
    view = HeadlessFigures(model.figures)
../bluesky-widgets/bluesky_widgets/headless/figures.py:40: in __init__
    self._add_figure(figure_spec)
../bluesky-widgets/bluesky_widgets/headless/figures.py:55: in _add_figure
    figure = HeadlessFigure(figure_spec)
../bluesky-widgets/bluesky_widgets/headless/figures.py:125: in __init__
    self._axes[axes_spec.uuid] = MatplotlibAxes(model=axes_spec, axes=axes)
../bluesky-widgets/bluesky_widgets/_matplotlib_axes.py:60: in __init__
    self._add_image(image_spec)
../bluesky-widgets/bluesky_widgets/_matplotlib_axes.py:131: in _add_image
    array = image_spec.func(run)
../bluesky-widgets/bluesky_widgets/models/plot_builders.py:567: in _transform
    (data,) = numpy.asarray(
../../../miniconda3/envs/py38/lib/python3.8/site-packages/numpy/core/_asarray.py:83: in asarray
    return array(a, dtype, copy=False, order=order)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/common.py:131: in __array__
    return np.asarray(self.values, dtype=dtype)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/dataarray.py:569: in values
    return self.variable.values
../../../miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/variable.py:510: in values
    return _as_array_or_item(self._data)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/variable.py:272: in _as_array_or_item
    data = np.asarray(data)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/numpy/core/_asarray.py:83: in asarray
    return array(a, dtype, copy=False, order=order)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/array/core.py:1376: in __array__
    x = self.compute()
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/base.py:167: in compute
    (result,) = compute(self, traverse=False, **kwargs)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/base.py:452: in compute
    results = schedule(dsk, keys, **kwargs)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/threaded.py:76: in get
    results = get_async(
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/local.py:486: in get_async
    raise_exception(exc, tb)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/local.py:316: in reraise
    raise exc
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/local.py:222: in execute_task
    result = _execute_task(task, data)
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/core.py:121: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/core.py:121: in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/core.py:121: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/core.py:121: in <genexpr>
    return func(*(_execute_task(a, cache) for a in args))
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/core.py:121: in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
../../../miniconda3/envs/py38/lib/python3.8/site-packages/dask/array/core.py:100: in getter
    c = a[b]
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
../../../miniconda3/envs/py38/lib/python3.8/site-packages/h5py/_hl/dataset.py:489: in __getitem__
    if is_empty_dataspace(self.id):
../../../miniconda3/envs/py38/lib/python3.8/site-packages/h5py/_hl/base.py:87: in is_empty_dataspace
    if obj.get_space().get_simple_extent_type() == h5s.NULL:
h5py/_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py/_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: Not a dataset (not a dataset)

h5py/h5d.pyx:289: ValueError
```

</details>

dylanmcreynolds commented:
Hmmm....nothing comes to mind. The test is yielding the file, and I'd expect to see a print in your output saying "closing file".

Sure it's an issue with file being closed and not just a h5py complaining that the dataset at the key being requested doesn't exist? I am not sure I see what key it would be in this code. I think my next step here would be to throw a print statement here: https://github.com/als-computing/splash-ingest/blob/05750bef15cad486c3787b44a426dc395c2aba4d/splash_ingest/handlers.py#L14 and see what key is being sent.

dylanmcreynolds commented:
FWIW, to get an image out of the primary stream in this test, I'd expect the key to be `:exchange:data`

danielballan commented:
Thanks. I put a breakpoint there, and I found that the `key` is as you say it should be. No problem there. If I override the `close()` method on the handler to do nothing (150d1b3) everything works and the tests pass.

The file will still be closed (via the garbage collector) but this is sloppy and I don't think it should be merged as is. Something is closing the handlers too fast. This is probably a bug somewhere in the bluesky stack that just happens to be being exercised here.

danielballan commented:
OK, I think the is in good shape now, other than the fact that it relies on an un-merged, un-released branch on bluesky-widgets.

I have re-instated `close()`. If the handler is _explicitly_ closed then it should clean up all of its internal open files (of course). But if the clean up is left to the garbage collector, cleaning up the handler shouldn't necessarily closes in open files if other objects (`dask.array`s in this case) are still holding references to those files.

This is fixed here by overriding `__del__` and will be fix upstream in https://github.com/bluesky/area-detector-handlers/pull/25. When that is released, patching `__del__` here we be unnecessary.

danielballan commented on splash_ingest/ingestors.py:
```diff
@@ -253,6 +262,44 @@ def calc_num_events(mapping_fields: List[StreamMappingField], file):
+        # public or we'll choose a different mechanism. - Dan Allan, Dec 2020
+
+        # The base class gives us a 2D image by slicing the middle of any
+        # higher dimensions.
```
Elaborating on this comment:

In the base class `__init__` we are effectively doing "select" (which columns) and "combine" (make an array of the right shape to plot). If you want to leave that logic as it is but add additional reduction / cleaning / whatever, you have to subclass override a method that I'm not even sure should exist. That feels extreme; there ought to be a gentler extension mechanism for something this simple.

The use case here might be described as inserting a step in the data processing pipeline on the way to plotting.

dylanmcreynolds replied:
After the event_model fix is released, will this need to be removed? Is the summary of that fix that is we never need to close h5 files explicitly? That when the file is GC'd, it gets closed auto-magically? Is the close method in h5py really just intended for ending write access?

danielballan replied:
Will continue this thread near the relevant lines.

danielballan replied:
Reflections on this:
1. I should add a first-class concept of (auto-)scaling to bluesky-widgets' `Image` plot builder. It was useful for Splash to point that out.
2. I think the right way to formulate it is a manipulating on the scaling (log vs linear) and range (min, max) of the color bar, not by changing the array itself. We should hand the actual data to matplotlib (or whatever plotting framework) so that when there is a colorbar, its scale represents physically-meaning values.
3. I still feel unsure about how best to make it easy to extend whatever built-in heuristics and/or reduction bluesky-widgets built-in classes do. I do not think subclassing is the way.

dylanmcreynolds replied:
Yeah, it seems to me that in some future version, it might be nice to think of handy reusable transform helpers that could be plugged in in spots like this. Heck, I don't even remember where I got the log_and_auto_contrast code originally. :)

dylanmcreynolds commented:
Given the timing of this, I'd like to hold off on merging until it can run on pip-installable versions of dependencies, even if those are pre-release.

dylanmcreynolds commented:
Build is failing because this depends on un-released projection code from databroker

danielballan commented on splash_ingest/handlers.py:
```diff
@@ -21,3 +21,9 @@ def __call__(self, point_number, key):
+
+    def __del__(self):
+        # Explicitly do NOT close on __del__.
+        # This will be fixed upstream in the base class (in PR
```
> After the event_model fix is released, will this need to be removed? Is the summary of that fix that is we never need to close h5 files explicitly? That when the file is GC'd, it gets closed auto-magically? Is the close method in h5py really just intended for ending write access?

I think you mean when `area-detector-handlers` is released. https://github.com/bluesky/area-detector-handlers/pull/25 No, this will become superfluous and _may_ be removed but it will do no harm if left as is. Yes, when GC-ed, the files are closed auto-magically. You could argue that `close()` is more important for write (because closing it makes it possible for another process to read from or write to that file). But I think `close()` still matters for read... it's just more convenient to rely on GC in situations where many different parts of the code are using the file, whereas in simple situations like

```py
with h5py.File(..., "r") as file:
    data = file[...]
# close() is called on exit from context
```

there no reason to rely on GC and it's "clean" to explicitly release the file.

dylanmcreynolds replied:
I am confused. If `HandlerBase.__del__` is not calling `close()` anymore, than where is it the file being closed?

danielballan replied:
It's being closed by the GC whenever the `h5py.File` itself goes out of scope. In this case, that will occur whenever the last `dask.array` that wraps a dataset from that file performs its I/O is replaced by the resultant `numpy.ndarray` within its `xarray.Dataset`.

danielballan replied:
> But I think close() still matters for read... it's just more convenient to rely on GC in situations where many different parts of the code are using the file, whereas in simple situations like

I have made muddied the waters here. In the abstract, I think `close()` is still important to implement and should be used in simple situations like the code snippet above, but this is not a simple situation, so we should rely on the GC.

danielballan commented:
Refactored to use `AutoPlotter` abstract base class, just released in `bluesky-widgets` v0.0.5. Rebased on `master` to resolve conflicts. Pending review, I think this is now in decent shape and ready to merge.

dylanmcreynolds commented on splash_ingest/ingestors.py:
```diff
@@ -253,6 +262,49 @@ def calc_num_events(mapping_fields: List[StreamMappingField], file):
+    return np.asarray(auto_contrast_image)
+
+
+class AutoThumbnailModel(AutoPlotter):
```
LOVE!

dylanmcreynolds commented on splash_ingest/ingestors.py:
```diff
@@ -95,18 +101,24 @@ def generate_docstream(self):
+        document_cache('resource', hd5_resource.resource_doc)
         yield 'resource', hd5_resource.resource_doc
 
         thumbnail_created = False  # for now, we'll just created one thumbnail per run, first 2d image we find
```
I guess thumbnail_created is no longer needed, its usage was removed.

dylanmcreynolds commented:
I tested this code with my standard test from tomography (a 3.3G file from the tomobank). The line 

`        model.add_run(bluesky_run)`

pulls in all of my RAM until the process finally crashes. This was not a problem with the prior thumbnail code.

dylanmcreynolds commented:
Poking around a little. The memory suck starts at this line:
https://github.com/bluesky/bluesky-widgets/blob/ecc1ee02d85858ca4d2976356bc9cb70a5e015d6/bluesky_widgets/models/plot_builders.py#L382

self.field is the lambda defined in this PRs ingestors.py:
```
images = Images(
                    field=lambda run: _log_and_auto_contrast(run[stream_name][field]),
                    needs_streams=(stream_name,))
```
That looks a little odd to me, that that is a lambda returning a nunpy array rather than a field name. Also, if it was a field name, shouldn't plot_builder be using `to_dask() `rather than ` read()`? 

Also, in ingestors.py, the parameter for that lambda is `run[stream_name][field]`, which doesn't evaluate (TypeError: 'BlueskyEventStream' object is not subscriptable). Since `_log_and_auto_contrast` expects an array, shouldn't it be something like `run[stream_name].to_dask()[field][0]`? There's probably something I don't understand going on.

danielballan commented:
Lots of good points here.

> https://github.com/bluesky/bluesky-widgets/blob/ecc1ee02d85858ca4d2976356bc9cb70a5e015d6/bluesky_widgets/models/plot_builders.py#L382

That should be using `to_dask()`. Does this help?

https://github.com/bluesky/bluesky-widgets/pull/89

> That looks a little odd to me, that that is a lambda returning a nunpy array rather than a field name.

The plot builder API accepts strings like

```py
Lines(x="log(temperature)", y="intensity / I0")
```

which is really fluid for simple expressions (and plays well with text input fields in views). But when the expressions become complex, stuffing them into a string is awkward and anti-helpful. Code linters and code formatters don't parse it. At certain point, it makes sense to switch to using a function. This is equivalent:

```py
Lines(x=lambda temperature: log(temperature), y=lambda intensity, I0: intensity / I0)
```

Those `lambda` functions are evaluated by inspecting their signatures and resolving the parameter names in the same namespace that we use for the string-based usage above. It's dependency injection via signature inspection. This verges on "too magical", but it's not a fairly common pattern, and in this situation it enables the user to access data by name without having the structure of a BlueskyRun in their head. That is, `lambda intensity, I0: intensity / I0` is more fluid than `run["primary"].read()["intensity"] / run["primary"].read()["I0"]`. The magic is in https://github.com/bluesky/bluesky-widgets/blob/ecc1ee02d85858ca4d2976356bc9cb70a5e015d6/bluesky_widgets/models/utils.py#L114-L163 Feel free to push back on this if you think it's headed in a bad direction. I feel decent about it based on how it feels to use so far, but it's not at all too late to pivot.

Reflecting on `Images(field=...)` in particular I think we should reconsider the name `field`, as that _strongly_ suggests a string field name, where `Lines(x=..., y=...)` are a little more generic.

> Also, the test function doesn't go through the Images._add_images code path. I'm guessing this is because the data set is 1D and making a plot instead of 2D and making an image?

I don't think it can be making a plot. This code doesn't do anything with `Lines`. It makes `Images` or nothing.

If your test function doesn't go through the `Images.add_images` code path, how can https://github.com/bluesky/bluesky-widgets/blob/ecc1ee02d85858ca4d2976356bc9cb70a5e015d6/bluesky_widgets/models/plot_builders.py#L382 be the issue?

dylanmcreynolds commented:
FWIW, I went into the file system to see the png that the unit test is creating...

![-1](https://user-images.githubusercontent.com/40469975/104392755-9fd31c00-54f7-11eb-8bd4-0dfc67dce09a.png)

dylanmcreynolds commented:
Thanks for all the description. Wow, python is cute.

I tried changing that call to to_dask(), but that instantly throws:

```
Traceback (most recent call last):
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/xarray/core/dataset.py", line 1259, in _construct_dataarray
    variable = self._variables[name]
KeyError: <function AutoThumbnailModel.handle_new_stream.<locals>.<lambda> at 0x7f6e9587fc10>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/bluesky_live/event.py", line 514, in _invoke_callback
    cb(event)
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/bluesky_widgets/models/plot_builders.py", line 382, in _add_images
    array_shape = run.primary.to_dask()[self.field].shape
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/xarray/core/dataset.py", line 1359, in __getitem__
    return self._construct_dataarray(key)
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/xarray/core/dataset.py", line 1261, in _construct_dataarray
    _, name, variable = _get_virtual_variable(
  File "/home/dylan/work/als-computing/splash-ingest/env/lib/python3.8/site-packages/xarray/core/dataset.py", line 156, in _get_virtual_variable
    raise KeyError(key)
KeyError: <function AutoThumbnailModel.handle_new_stream.<locals>.<lambda> at 0x7f6e9587fc10>
```

danielballan commented:
TO DO:

- [ ] Release event-model with https://github.com/bluesky/event-model/pull/204 and pin that minimum version here.
- [ ] Experiment with a PIL-based view instead of this matplotlib-based one. This will (1) produce the desired result---an image with no colorbar or axes labels---and (2) probably run faster.

danielballan commented:
Rebased on master. The fix to event-model has been released in v1.17.1 and I have updated the minimum version pin. Next (final?) step before merging is to make a PIL-based view derived from the matplotlib-based `HeadlessFigures` view, and to drop the matplotlib dependency.

dylanmcreynolds commented:
I love this idea. But the truth is that ingest has been refactored a lot, and I think there needs more in bluesky widgets. I'd like to revisit this at a future date. Closing for now.
